Create analyzer for non-serializable TheoryData type arguments (xunit/xunit#2866)

jared-chevalier opened the pull request:
Suggested initial version of a solution for issue [xunit/xunit#2866](https://github.com/xunit/xunit/issues/2866).

Includes analyzer for proposed new rule xUnit1044: TheoryData type arguments should be serializable.

Please let me know if you have any feedback.

# Solution

## `Serializability` enumeration

- `NeverSerializable`
- `PossiblySerializable`
- `AlwaysSerializable`

## `TheoryDataTypeArgumentFinder` class

- Finds all `TheoryData<>` type arguments for a data source referenced by a given `[ClassData]` or `[MemberData]` attribute in a given test class, if applicable
- If the data source's type is not compatible with a `TheoryData<>` class (such as if it is a member with the type `IEnumerable<object[]>`), then no type arguments will be found
- If the data source's type derives from a `TheoryData<>` base class, then the type arguments for the base class will be found 

## `SerializabilityAnalyzer` class

- Analyzes a given type to determine whether it is always, possibly, or never serializable
- The static analysis of a type's serializability closely corresponds to the logic found in `SerializationHelper` and `MemberDataAttributeBase` from the `xunit` repository
- See discussion in issue [xunit/xunit#2866](https://github.com/xunit/xunit/issues/2866)

## `TypeSymbols` class

- Lazy caching wrapper around `TypeSymbolFactory` (plus the `Compilation` and `XunitContext`), for reusing already generated type symbols

## `TheoryDataTypeArgumentsShouldBeSerializable` analyzer

- Analyzes `MethodDeclarationSyntax` nodes with a `[Theory]` attribute
- For each `[ClassData]` or `[MemberData]` attribute: Uses `TheoryDataTypeArgumentFinder` to find all `TheoryData<>` type arguments for the referenced class or member, if applicable
- For each type argument found: Uses `SerializabilityAnalyzer` to check whether the type is serializable (unless the type should be ignored)
- For each never serializable type argument: Reports an xUnit1044 diagnostic for the data attribute syntax node, with a strongly phrased message: `The type argument {0} is not serializable. Consider using a type that is known to be serializable.`
- For each possibly serializable type argument: Reports an xUnit1044 diagnostic for the data attribute syntax node, with a weakly phrased message: `The type argument {0} might not be serializable. Consider using a type that is known to be serializable.`

# Key points

## Diagnostic severity

- The diagnostic currently has `Info` severity. Would `Warning` be more appropriate?
- Currently, using `Warning` would break the build because an existing test class, `SetEqualityAnalyzerTests`, uses a tuple type argument for a `MatrixTheoryData` data source, and warnings are treated as errors
- A solution would be to skip analyzing data sources associated with `DisableDiscoveryEnumeration = true`, but that has not been implemented yet (see discussion in issue [xunit/xunit#2866](https://github.com/xunit/xunit/issues/2866))

## Statically analyzing serializability of enumerations

- See discussion in issue [xunit/xunit#2866](https://github.com/xunit/xunit/issues/2866)
- Any given enumeration type is either always serializable (from a local assembly) or never serializable (from the GAC)
- As far as I can tell, static analysis cannot determine whether an enumeration type is from a local assembly or from the GAC
- As a result, all enumeration types are treated as possibly serializable
- However, the analyzer ignores them, in order to prevent a diagnostic from being found for every single enumeration type

## IXunitSerializable type symbol

- `IXunitSerializable` is in namespace `Xunit.Abstractions` in xUnit version 2, but in `Xunit.Sdk` in version 3
- The version 2 type symbol is accessible via a lazy cache in `xunitContext.V2Abstractions?.IXunitSerializableType` (as well as directly via `TypeSymbolFactory`)
- There does not seem to be any corresponding property in any of the version 3 contexts. Would it make sense to add one somewhere in the contexts? Ideally there would be a common interface, implemented for each version, that provided the type symbol for the current version
- I added the `TypeSymbolFactory.IXunitSerializable_V3` factory method for the version 3 type symbol
- My `TypeSymbols` class currently uses the following logic to access the appropriate type symbol: `TypeSymbolFactory.IXunitSerializable_V3(compilation) ?? xunitContext.V2Abstractions?.IXunitSerializableType`

jared-chevalier committed:
Extract IsInstanceOf extension method from ContainsAttributeType

jared-chevalier committed:
Add UnwrapNullable extension method

jared-chevalier committed:
Move and format TheoryDataN method for style consistency

jared-chevalier committed:
Add additional factory methods to TypeSymbolFactory

jared-chevalier committed:
Implement initial version of xUnit1044 analyzer

jared-chevalier committed:
Add test ensuring no xUnit1044 diagnostic for non-TheoryData data source

bradwilson commented on src/xunit.analyzers.tests/Analyzers/X1000/TheoryDataTypeArgumentsShouldBeSerializableTests.cs:
```diff
@@ -0,0 +1,426 @@
+	[MemberData(nameof(TheoryDataMembers), "TimeOnly[]")]
+	[MemberData(nameof(TheoryDataMembers), "TimeOnly?")]
+	[MemberData(nameof(TheoryDataMembers), "TimeOnly?[]")]
+	public async void GivenTheoryMethod_WithDateOnlyOrTimeOnlyInTheoryDataMember_DoesNotFindDiagnostic(
```
You could collapse this into `GivenTheoryMethod_WithSerializableTheoryDataMember_DoesNotFindDiagnostic` and just put the member data itself inside the `#if` block.

jared-chevalier replied:
Good point.

bradwilson commented on src/xunit.analyzers/Utility/Descriptors.xUnit1xxx.cs:
```diff
@@ -402,7 +402,14 @@ public static partial class Descriptors
 		);
 
-	// Placeholder for rule X1044
+	public static DiagnosticDescriptor X1044_TheoryDataTypeArgumentsShouldBeSerializable { get; } =
```
As discussed in the issue thread, I believe it would be best to emit two different diagnostics here, so users can turn the "maybe" one off without having to turn the "definitely" one off.

jared-chevalier replied:
Sounds good.

bradwilson commented on src/xunit.analyzers/Utility/Descriptors.xUnit1xxx.cs:
```diff
@@ -402,7 +402,14 @@ public static partial class Descriptors
+			"TheoryData type arguments should be serializable",
+			Usage,
+			Info,
+			"The type argument {0} {1}. Consider using a type that is known to be serializable."
```
The two diagnostics should get the updated language from the issue thread.

jared-chevalier replied:
Sounds good.

bradwilson commented on src/xunit.analyzers/X1000/TheoryDataTypeArgumentsShouldBeSerializable.cs:
```diff
@@ -0,0 +1,567 @@
+	{
+		Guard.ArgumentNotNull(context);
+		Guard.ArgumentNotNull(xunitContext);
+
```
We can skip early for any project which isn't targeting .NET Framework.

jared-chevalier replied:
Can you explain this? I don't understand the reasoning here.

bradwilson replied:
The GAC only exists in .NET Framework. It doesn't exist for .NET Core/.NET.

bradwilson replied:
I realize now I really put this comment in the wrong place and with nowhere near enough words. 😝 We can skip "Enum in the GAC" analysis when the context is analyzing a .NET Framework targeting project.

jared-chevalier replied:
Ok, that makes more sense! 😂

jared-chevalier replied:
Could you clarify how to determine whether the context is analyzing a project targeting .NET Framework?

bradwilson replied:
I've never done it myself, so I don't know. I assume it would be part of the context that's given, but I'd have to poke around it in a debugger to figure out how (and whether it's even possible), which is how I do 100% of my Roslyn work, sadly. 😂

bradwilson replied:
Since you don't know, we can leave it off for now and I'll try to add it later.

jared-chevalier replied:
> which is how I do 100% of my Roslyn work, sadly.

😂

I poked around the context aggregate, and briefly searched online, but didn't find anything.

I'll spend a little time later investigating how feasible it would be to check whether Roslyn has source-level access to an enum definition, like you suggested.

If it's not, and if we can't figure out how to check the target framework, then I suppose the enum serializability problem is intractable after all. In which case, the analyzer should presumably continue ignoring all enum types? Since otherwise, every single enum type would produce the "might not be serializable" diagnostic, which I imagine would be more bothersome than useful.

bradwilson replied:
> I'll spend a little time later investigating how feasible it would be to check whether Roslyn has source-level access to an enum definition, like you suggested.

I was able to dig this out in the debugger. 😁 Looks like you can look for `ISourceAssemblySymbol` on the type's `ContainingAssembly`. Here's one where the type is local:

```
> type.ContainingAssembly is ISourceAssemblySymbol
true
```

And here's one where it isn't:

```
> context.Compilation.GetTypeByMetadataName("System.ConsoleColor").ContainingAssembly is ISourceAssemblySymbol
false
```

The actual concrete types were helpfully named `SourceAssemblySymbol` and `NonSourceAssemblySymbol` 😆 

Still looking for access to target framework...

bradwilson replied:
I think the more I think about it, we're only circling around to something vaguely successful for enums, so we should just ignore enums entirely. The GAC requirement isn't reliably testable, and honestly most people these days are probably targeting .NET (and not .NET Framework), so let's just let the .NET Framework people fend for themselves, which is exactly where they would've been before this analyzer.

jared-chevalier replied:
> The actual concrete types were helpfully named `SourceAssemblySymbol` and `NonSourceAssemblySymbol` 😆

Thanks, good to know!

> I think the more I think about it, we're only circling around to something vaguely successful for enums, so we should just ignore enums entirely. The GAC requirement isn't reliably testable, and honestly most people these days are probably targeting .NET (and not .NET Framework), so let's just let the .NET Framework people fend for themselves, which is exactly where they would've been before this analyzer.

Alright, sounds good.

bradwilson commented on src/xunit.analyzers/X1000/TheoryDataTypeArgumentsShouldBeSerializable.cs:
```diff
@@ -0,0 +1,567 @@
+			AttributeData dataAttribute,
+			INamedTypeSymbol testClass)
+		{
+			if (dataAttribute.IsInstanceOf(typeSymbols.ClassDataAttribute))
```
This should be an exact type match, since any class derived from `ClassDataAttribute` does not necessarily have to have a constructor of this shape.

jared-chevalier replied:
Good point!

bradwilson commented on src/xunit.analyzers/X1000/TheoryDataTypeArgumentsShouldBeSerializable.cs:
```diff
@@ -0,0 +1,567 @@
+			if (dataAttribute.IsInstanceOf(typeSymbols.ClassDataAttribute))
+				return dataAttribute.ConstructorArguments.FirstOrDefault().Value as INamedTypeSymbol;
+
+			if (dataAttribute.IsInstanceOf(typeSymbols.MemberDataAttribute))
```
Even though `MemberDataAttribute` is currently sealed, philosophically I'd still like to see an exact match here as well. If it turns out we decide to unseal it for v3, I'd want this code to continue to behave in the correct way; this is not currently planned, but I'd rather be safe. 😄

jared-chevalier replied:
No problem, makes sense!

bradwilson reviewed (changes requested):
Looking very good. Just a few comments based on the thread in the associated issue.

jared-chevalier committed:
Merge theory data rows

Testing for both non-parameterized and parameterized methods as TheoryData<> members did not end up providing much value, so merging them seems to be the better option. Merging them will also significantly reduce the number of test cases, improving test performance.

jared-chevalier committed:
Merge test methods

Move DateOnly/TimeOnly test cases, including conditional compilation directives, from separate test method GivenTheoryMethod_WithDateOnlyOrTimeOnlyInTheoryDataMember_DoesNotFindDiagnostic to the main test method for serializable theory data members, GivenTheoryMethod_WithSerializableTheoryDataMember_DoesNotFindDiagnostic, eliminating the separate test method.

jared-chevalier committed:
Split diagnostic X1044 into separate diagnostics X1044 and X1045

X1044: TheoryData type argument is not serializable.
X1045: TheoryData type argument might not be serializable.

jared-chevalier committed:
Group tests in nested classes by diagnostic found

jared-chevalier committed:
Check for exact match of data attribute type

ClassDataAttribute is not sealed, and any class derived from it does not necessarily have to have a constructor of this shape.

Even though MemberDataAttribute is currently sealed, if it turns out we decide to unseal it for version 3, this code should continue to behave in the correct way.

jared-chevalier committed:
Ignore theory methods with discovery enumeration disabled

jared-chevalier committed:
Shorten test method names for improved readability

jared-chevalier committed:
Ensure TheoryData class is ignored if discovery enumeration is disabled

jared-chevalier commented on src/xunit.analyzers/Utility/Descriptors.xUnit1xxx.cs:
```diff
@@ -402,9 +402,23 @@ public static partial class Descriptors
 
-	// Placeholder for rule X1044
+	public static DiagnosticDescriptor X1044_TheoryDataTypeArgumentsShouldBeSerializable { get; } =
+		Diagnostic(
+			"xUnit1044",
+			"TheoryData type arguments should be serializable",
+			Usage,
+			Info,
+			"The type argument {0} is not serializable, which will cause Test Explorer to not enumerate individual data rows. Consider using a type that is known to be serializable."
+		);
 
-	// Placeholder for rule X1045
+	public static DiagnosticDescriptor X1045_TheoryDataTypeArgumentsShouldBeDefinitelySerializable { get; } =
+		Diagnostic(
+			"xUnit1045",
+			"TheoryData type arguments should be definitely serializable",
+			Usage,
+			Info,
+			"The type argument {0} might not be serializable, which may cause Test Explorer to not enumerate individual data rows. Consider using a type that is known to be serializable."
+		);
```
Do you have any alternative suggestions for titles and property names for these two diagnostics?

I tried to come up with a pair of titles that clearly distinguished between the two without excessive wordiness or double negatives. This is what I came up with after trying out some options, but I'm not sure I'm happy with it.

bradwilson replied:
Yeah, this one's weird. It's almost like it's better to invert the naming to be the problem rather than the solution because "Should" is applies to both.

Maybe:

`X1044_TheoryDataTypeArgumentMightNotBeSerializable`
`X1045_TheoryDataTypeArgumentIsNotSerializable`

With titles:

`TheoryData type arguments might not be serializable`
`TheoryData type arguments are not serializable`

Thoughts?

jared-chevalier replied:
I went through the existing diagnostics and did an inventory of the language used, to see whether there is a precedent for this.

Most use standard prescriptive language: `must (not)`, `should (not)`, `can(not)`, `is recommended`.

A few use imperative voice: `ensure`, `do not`, `use`.

Only one uses descriptive language in the property name, describing the problem rather than prescribing the solution: `X1028_TestMethodHasInvalidReturnType` (though the title is phrased prescriptively: `Test method must have valid return type`).

I like your suggestions, as they are clear and simple.

Here's another option, using imperative voice with `avoid using`:

`Avoid using TheoryData type arguments that are not serializable`
`Avoid using TheoryData type arguments that might not be serializable`

I suppose mine are slightly more consistent with existing style, but also slightly more wordy. I'm happy to go with whichever you prefer.

jared-chevalier replied:
By the way, did you intend to switch the diagnostic ids? I currently have X1044 as the "is not serializable" diagnostic and X1045 as the "might not be serializable" one.

bradwilson replied:
No, that was unintentional.

bradwilson replied:
I like your suggestions. 👍🏼

jared-chevalier replied:
Ok, sounds good.

jared-chevalier committed:
Update SerializabilityAnalyzer documentation regarding enumerations

jared-chevalier committed:
Rename diagnostics X1044 and X1045 for clarity

bradwilson reviewed (approved):
Awesome job! Thanks

bradwilson commented:
I decided to take this work and make parallel analyzers for v3's `TheoryDataRow` constructor: https://github.com/xunit/xunit.analyzers/commit/68525f89d851fb482221f3ffdf0560929a9be30d

jared-chevalier commented:
> Awesome job! Thanks

You're welcome, it was my pleasure!

> I decided to take this work and make parallel analyzers for v3's `TheoryDataRow` constructor: [68525f8](https://github.com/xunit/xunit.analyzers/commit/68525f89d851fb482221f3ffdf0560929a9be30d)

Nice!
